libraries/lib-ipc/IPCServer.cpp

```diff
@@ -0,0 +1,116 @@
+#include "IPCServer.h"
+#include "IPCChannel.h"
+
+#include <atomic>
+#include <thread>
+
+#include "internal/ipc-types.h"
+#include "internal/socket_guard.h"
+#include "internal/BufferedIPCChannel.h"
+
+class IPCServer::Impl
+{
+   std::atomic<bool> mAlive{true};
+   std::unique_ptr<std::thread> mWorker;
+public:
+
+   ~Impl()
+   {
+      mAlive = false;
+      if(mWorker)
+         mWorker->join();
+   }
+
+   bool CreateChannel(int id, IPCChannelStatusCallback* callback)
+   {
+      if(mWorker)
+         return false;
+
+      auto serverfd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+      if(serverfd == INVALID_SOCKET)
+         return false;
+
+      socket_guard serverfd_guard { serverfd };
+
+      sockaddr_in addrin {};
+      addrin.sin_family = AF_INET;
+      addrin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+      addrin.sin_port = htons(static_cast<u_short>(id));
+
+      if(bind(serverfd, reinterpret_cast<const sockaddr*>(&addrin), sizeof(addrin)) == SOCKET_ERROR)
+         return false;
+
+      if(listen(serverfd, 1) == SOCKET_ERROR)
+         return false;
+
+      mWorker = std::make_unique<std::thread>([this, id, callback, serverfd, guard = std::move(serverfd_guard)]()
```

Comment:
If channel becomes a reference instead of pointer, remember to pass `std::ref(channel)` to avoid the copy constructor.

Reply:
Argument `id` is not used